### PR TITLE
Tenant deletion when using blocks storage

### PR DESCRIPTION
**Pull request by pstibrany**
**What this PR does**: This PR introduces new API for tenant deletion when using blocks storage. There are two API endpoints: one to trigger tenant deletion (it really only writes mark file with tenant prefix), another one to check for status (= check for blocks).

In this PR ingester checks for tenant deletion mark, and 1) stops shipping blocks for such tenant, and 2) closes TSDB faster (if closing is enabled).

In subsequent PRs, more components will react on deletion mark:
- compactor will clean up blocks in the bucket
- ruler will clean up rule groups, and stop rule evaluation [ruler will use separate marker in its own bucket]
- alertmanager will clean up its per-tenant config and state [alertmanager will use separate marker in its own bucket]

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/storage/tsdb/tenant_deletion_mark.go
Comment (pracucci) [inline pkg/storage/tsdb/tenant_deletion_mark.go:14]: Two things:
1. Can we store it under the `marker/` prefix, within the tenant location? This way will work nicely with the bucket index too (we'll store the block deletion marks there).
2. What do you think to name it `tenant-deletion-mark.json` to keep naming consistent with the block deletion mark?

File: pkg/storage/tsdb/tenant_deletion_mark.go
Comment (pracucci) [inline pkg/storage/tsdb/tenant_deletion_mark.go:18]: I would keep it consistent with the block deletion mark:

```
	DeletionTime int64 `json:"deletion_time"`
```

File: pkg/storage/tsdb/tenant_deletion_mark.go
Comment (pracucci) [inline pkg/storage/tsdb/tenant_deletion_mark.go:38]:
```suggestion
	return errors.Wrap(bkt.Upload(ctx, markerFile, bytes.NewReader(data)), "upload tenant deletion mark")
```

File: pkg/storage/tsdb/tenant_deletion_mark.go
Comment (pracucci) [inline pkg/storage/tsdb/tenant_deletion_mark.go:34]:
```suggestion
		return errors.Wrap(err, "serialize tenant deletion mark")
```

File: pkg/chunk/purger/blocks_purger_api.go
Comment (pracucci) [inline pkg/chunk/purger/blocks_purger_api.go:40]: We return `http.StatusUnauthorized` is other APIs. We should be consistent here too.

_Same comment applies to `DeleteTenantStatus()`._

File: pkg/chunk/purger/blocks_purger_api.go
Comment (pracucci) [inline pkg/chunk/purger/blocks_purger_api.go:44]: Should we check if already exists, in order to not overwrite it? May be useful to not overwrite, to make sure that the "deletion time" in the marker is the oldest one.

**Review by pracucci — COMMENTED** [review]
Very good job! Design and implementation LGTM. I just left few minor comments, some related to my work on bucketindex.

Have you considered:
- Add a CHANGELOG entry
- Update the HTTP API doc (mentioning it's experimental)
- Add the tenant deletion to the list of experimental features in the doc

File: pkg/ingester/ingester_v2.go
Comment (pstibrany) [reply pkg/ingester/ingester_v2.go]: I've changed that since it makes tests that have no bucket defined to fail.

File: pkg/ingester/ingester_v2.go
Comment (pstibrany) [reply pkg/ingester/ingester_v2.go:1487]: I don't think we should. Supporting rollback is entire new can of worms (compactor will actually delete data when it finds the mark, it cannot do undelete), and I don't think we should give false hope here by setting it back to false. Furthermore, once `deletionMarkFound` is set to true, we don't even get to this point.

File: pkg/storage/tsdb/tenant_deletion_mark.go
Comment (pstibrany) [reply pkg/storage/tsdb/tenant_deletion_mark.go:14]: Sure.

File: pkg/chunk/purger/blocks_purger_api.go
Comment (pstibrany) [reply pkg/chunk/purger/blocks_purger_api.go:40]: I don't think this is correct. This endpoint uses authentication middleware, which already checks if there is correct x-scope-orgID header, and stores tenant into context. Auth middleware already returns `http.StatusUnauthorized`. If we get error here, it can be for different reasons, e.g. using multi-tenants.

Comment (pstibrany) [reply pkg/chunk/purger/blocks_purger_api.go:40]: (Btw, I purger's other handler uses `http.StatusBadRequest`, instead of `http.StatusUnauthorized`. I copied it from there. I think it's more correct.)

File: pkg/chunk/purger/blocks_purger_api.go
Comment (pstibrany) [reply pkg/chunk/purger/blocks_purger_api.go:44]: It doesn't make difference from API point of view, and overwriting with more recent timestamp will not do any harm. Additionally, it would be best-effort-attempt only. I suggest that we instead log a message on successful creation, so that it is possible to find out if there were multiple markers, if needed for troubleshooting.

File: pkg/chunk/purger/blocks_purger_api.go
Comment (pracucci) [reply pkg/chunk/purger/blocks_purger_api.go:42]: I see your point. I'm OK keeping it as is in this PR. However, I think it's worth opening a GitHub issue about it, cause we should keep a keep a consistent approach within Cortex.

File: pkg/chunk/purger/blocks_purger_api.go
Comment (pracucci) [reply pkg/chunk/purger/blocks_purger_api.go:46]: It's true doesn't cause any harm and it's true would be a best-effort attempt, but I still believe keeping the original deletion timestamp in the marker would be useful for debugging (eg. you're assuming every Cortex users collects and stores Cortex logs for a reasonable amount of time, which I'm not sure is true).

Reversing your answer, I may say that even checking for existence and returning a "no content" status code without overwriting it doesn't cause any harm and would avoid to overwrite it under normal circumnstances.

In any case, not a blocker to me.

File: pkg/chunk/purger/blocks_purger_api.go
Comment (pracucci) [inline pkg/chunk/purger/blocks_purger_api.go:54]: Looks a bit weird using `http.Error()` to report a success. Maybe we could just return a 200 with empty response body (all in all the message doesn't bring any value) or, if we wanna keep the message, create an `util.WriteTextResponse()` similar to the already existing `util.WriteJSONResponse()`?

**Review by pracucci — APPROVED** [review]
Thanks for addressing my comments! LGTM. I left few more comments, mostly nits, but please take a look before merging.

File: pkg/storage/tsdb/tenant_deletion_mark.go
Comment (pstibrany) [reply pkg/storage/tsdb/tenant_deletion_mark.go:14]: >  I don't remember if it's safe with regards of filesystem-based bucket.

If it's not, I'd argue that filesystem-bucket is broken and should be fixed. Since we want to use `markers` prefix, we need to use some separator, and we already use `/` elsewhere. (E.g. `UserBucketClient` uses it in its `fullName` method)

Also renamed "filename" to path.

File: pkg/api/api.go
Comment (jtlisi) [inline pkg/api/api.go:236]: **Thought:** It's not an official convention, however, requests that are scoped to a particular tenant universally have a `/api/v1` prefix. The only exception I know of is the Alertmanager UI which is a bit of an edge case. It might be worth following/codifying that convention in this PR. However, I can see an argument to be made that this is a bit different since it eliminates the tenant from existence.

```suggestion
	a.RegisterRoute("/api/v1/delete_tenant", http.HandlerFunc(api.DeleteTenant), true, "POST")
	a.RegisterRoute("/api/v1/delete_tenant_status", http.HandlerFunc(api.DeleteTenantStatus), true, "GET")
```

File: pkg/chunk/purger/blocks_purger_api.go
Comment (jtlisi) [inline pkg/chunk/purger/blocks_purger_api.go:61]:
```
	RuleGroupsDeleted         bool   `json:"rule_groups_deleted"`
	AlertManagerConfigDeleted bool   `json:"alert_manager_config_deleted"`
```

I don't think these two fields are used anywhere?

**Review by jtlisi — APPROVED** [review]
LGTM, modulo a few non-blocking nits

File: pkg/api/api.go
Comment (pstibrany) [reply pkg/api/api.go:236]: I am also conflicted about this. On one hand, this is part of API, and we will keep it stable once development has finished.

On the other hand, it's part of "admin" API (eg. like /shutdown), and don't want to send wrong signal by putting it under `/api/v1` prefix. Someone may accidentally expose entire `/api/v1/*` to the users, which in case of `delete_tenant` would be pretty bad.

I'd suggest to keep it as is for now, we can always decide on supporting different path in the future.

File: pkg/chunk/purger/blocks_purger_api.go
Comment (pstibrany) [reply pkg/chunk/purger/blocks_purger_api.go:61]: You're right. For now they are just an example. More PRs are coming with support for deleting tenant's ruler and alertmanager data (config, state), and then they will be used as well (perhaps in a changed form).

File: pkg/ingester/ingester_v2.go
Comment (pracucci) [inline pkg/ingester/ingester_v2.go:274]: > Reports tsdbIdle if TSDB can be closed

Can also be closed on `tsdbTenantMarkedForDeletion`. Maybe we should just generalise the comment and say that "Checks whether the TSDN can be closed.".

File: pkg/ingester/ingester_v2.go
Comment (pracucci) [inline pkg/ingester/ingester_v2.go:1495]:
```suggestion
				level.Warn(util.Logger).Log("msg", "failed to check for tenant deletion mark before shipping blocks", "user", userID, "err", err)
```

File: pkg/ingester/ingester_v2.go
Comment (pracucci) [inline pkg/ingester/ingester_v2.go:1669]: The log messages ends with `"dir"` but the dif value is missing.

```suggestion
		level.Info(util.Logger).Log("msg", "deleted local TSDB, user marked for deletion", "user", userID, "dir", dir)
```

**Review by pracucci — APPROVED** [review]
Thanks for addressing my feedback. LGTM! I left few final nits, but no need for having me re-review it again. Thanks! 👏

**Pull request merged**
